### ISSUE TITLE
Zone Edge Effects in subchandra Castro Runs

### ISSUE THREAD

**Issue by abigailbishop** [AMReX-Astro/Castro#550]
When running the model in `Castro/Exec/science/subchandra` on Edison at Nersc and our computers at Stony Brook, the plots of `magvel` and `eint_E` (aka `eint_e`) show inexplicable edge effects on zones. Please see the attached two images of how this runs on Bender at Stony Brook. The `magvel` plot is on a log scale to better show these effects. They begin as soon as the simulation takes place, allowing them to escalate to dramatic effects with more plot files. 

To run this simulation, make sure `inputs_2d` has `amr.plot_int = 1` to plot every plot file, compile the simulation in `Castro/Exec/science/subchandra` via `make`, then run the executable via `./Castro2d.gnu.MPI.ex inputs_2d`. After 2 plot files, the effects become apparent on the `magvel` and `eint_e` plots (you may have to set them to log scale). 

![subch_plt00002_eint_e_slice](https://user-images.githubusercontent.com/20175394/54222556-089c7800-44cc-11e9-8f06-d564a9bd3d06.png)
![subch_plt00002_magvel_slice](https://user-images.githubusercontent.com/20175394/54222557-089c7800-44cc-11e9-9826-2f84a0ca92db.png)

**Comment by zingale** [comment]
Max tracked this down to the lack of an _error buffer_ at jumps in refinement.  The fix, which I will check in to the `subchandra/` problem is to set
```
amr.n_error_buf     = 2 2 2 2
```
in the inputs file (it is currently `0 0 0 0`.

**Comment by zingale** [comment]
@abigailbishop can you update, and make sure your inputs has this change and start up a new run?

**Comment by abigailbishop** [comment]
For sure! I'm running it now on Edison and Bender and will update with results soon.

**Comment by abigailbishop** [comment]
That didn't work so @zingale is suggesting adding `castro.update_sources_after_reflux = 0` to `inputs_2d` since he noticed these effects happen at jumps in refinement. He suspects that it may be from a reflux which verifies conservations of certain variables across the jump. 

Currently running this on Bender.

**Comment by abigailbishop** [comment]
On Bender we saw improvements by adding `castro.update_sources_after_reflux = 0` to the inputs file (images below). I've submitted jobs to Edison to see the long term difference between a run with `castro.update_sources_after_reflux = 0` in the inputs file and without. 

![subch_plt00001_eint_e_slice_withoutReflux](https://user-images.githubusercontent.com/20175394/54450471-d5095a00-4726-11e9-8bdd-2373704ab3cf.png) 
Shows `eint_E` in the simulation after 1 time step _without_ the reflux added to the inputs file. 

![subch_plt00001_eint_e_slice_withReflux](https://user-images.githubusercontent.com/20175394/54450533-fcf8bd80-4726-11e9-9cd2-c9ca24c98f9e.png)
Shows `eint_E` in the simulation after 1 time step _with_ the reflux added to the inputs file. There are faint signatures of effects at (r=3E8, z=0.2E9)  and again at (5.5E8, 1.1E9).

**Comment by maxpkatz** [comment]
Do you see this effect with castro.do_react = 0?

**Comment by abigailbishop** [comment]
It's currently set to 1, but I just kicked off a run with `castro.do_react = 0` and `castro.update_sources_after_reflux = 0`.

**Comment by abigailbishop** [comment]
TLDR: Adding `castro.update_sources_after_reflux = 0` and updating to `castro.do_react = 0` in the inputs file seems to resolve the issue.

Below is an image of the first plot file of a simulation _without_ `castro.update_sources_after_reflux = 0` and `castro.do_react = 1`

![subch_plt00001_eint_e_slice_withoutReflux](https://user-images.githubusercontent.com/20175394/55647284-5181dc80-57ab-11e9-9f2e-5e78301ae435.png)

Below is an image of the first plot file with `castro.update_sources_after_reflux = 0` and `castro.do_react = 1`.

![subch_plt00001_eint_e_slice_withReflux](https://user-images.githubusercontent.com/20175394/55647355-77a77c80-57ab-11e9-86b6-baeaf272d645.png)

Lastly, below is an image of the first plot file with `castro.update_sources_after_reflux = 0` and `castro.do_react = 0`.

The simulations of all scenarios have gotten farther than before, up to about plot file 85,000, before quitting where the first scenario gave the error of low pressure and rho_e we think is associate with the reflux problem and the second scenario had the excuse of a seg fault. The third quit at 15,000 because that what the inputs file said to do, and I have since increased that and am running it with the hopes that it runs far.
Also, now that I've updated Castro, Microphysics, and Amrex (which are all on the development branch), they are all running the same simulation.

**Comment by zingale** [comment]
can you also try running with:

`gravity.gravity_type = PoissonGrav` instead of `MonopoleGrav`?

**Comment by abigailbishop** [comment]
It looks like `gravity.gravity_type = PoissonGrav` coupled with the `castro.update_sources_after_reflux = 0` and `castro.do_react = 0` causes goofy trends. Here's the `eint_e` plot for the first plot file:

![subch_plt00001_eint_e_slicesubch_PoissonGrav](https://user-images.githubusercontent.com/20175394/56061425-632d2c00-5d37-11e9-8576-6115725c9876.png)

**Comment by zingale** [comment]
the issue appears to be due to `castro.small_dens` being the same as the lowest density in the initial model.  So we are seeing the effect of density resets.  Not sure why they are so dramatic though.  In my runs (with only 2 levels of refinement), if I set `castro.small_dens = 1.e-5`, then all the funniness went away.

**Comment by abigailbishop** [comment]
The `Castro.small_dens` along with `castro.update_sources_after_reflux = 0` have resolved the issue! A current run has made it 45,000 time steps without any weird effects on the magvel plot.
